lib/async_hooks.js

```diff
@@ -197,6 +198,89 @@ class AsyncResource {
   }
 }
 
+function getResource(id) {
+  const weak = idResourcePairing.get(id);
+  if (!weak) {
+    return undefined;
+  }
+  return weak.get();
+}
+
+function getCurrentResource() {
+  const id = executionAsyncId();
+  return getResource(id);
+}
+
+const idResourcePairing = new Map();
+const contexts = new Set();
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // Save new resource.
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResource = getResource(triggerAsyncId);
+    if (triggerResource) {
+      contexts.forEach((context) => {
+        context._propagate(resource, triggerResource);
+      });
+    }
+  },
+  destroy(asyncId) {
+    idResourcePairing.delete(asyncId);
+  }
+});
+
+const kResourceStore = Symbol('kResourceStore');
```

Comment:
A global store symbol won't work for obvious reasons: different `AsyncContext`s/`AsyncContext.run`s will be overwriting each others store. It has to be a separate symbol per `AsyncContext`, or even per `.run`.

Looks like current tests are not enough as they didn't fail once `WeakMap` was removed recently.

Reply:
Indeed Thank You for mentioning this exact line! This can be the the point I was trying to find why contexts are messing up in a userland.

Reply:
Good catch. I will change that